In [12]:
!pwd

/home/ubuntu/notebooks


In [14]:
!ls -a

 .		      'Class 3'
 ..		      'DNSC Assignment_1_6305'
 .ipynb_checkpoints    DNSC6305_Assignment1_Priyanka_Bhatia.ipynb
 AssignementQ3.ipynb   Untitled.ipynb
'Class 1'	       hamlet.txt
'Class 2'	       hound.txt


In [15]:
!whoami

ubuntu


In [16]:
!date

Fri Sep 23 05:50:54 UTC 2022


In [17]:
import os
os.listdir()

['Class 1',
 'hound.txt',
 'DNSC6305_Assignment1_Priyanka_Bhatia.ipynb',
 'Class 3',
 'hamlet.txt',
 'AssignementQ3.ipynb',
 'Class 2',
 'Untitled.ipynb',
 'DNSC Assignment_1_6305',
 '.ipynb_checkpoints']

### Question 1 Part A. Characters in The Hound of the Baskervilles 
Use the text available at https://s3.amazonaws.com/dmfa-2020/project-1/hound.txt for this part.
How many times are each of the following characters mentioned by name in the text of The
Hound of the Baskervilles?
Holmes, Watson, Barrymore, Mortimer

In [6]:
!wget https://s3.amazonaws.com/dmfa-2020/project-1/hound.txt

--2022-09-19 21:22:12--  https://s3.amazonaws.com/dmfa-2020/project-1/hound.txt
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.217.203.88
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.217.203.88|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 345659 (338K) [text/plain]
Saving to: ‘hound.txt’

hound.txt           100%[===================>] 337.56K  --.-KB/s    in 0.008s  

2022-09-19 21:22:12 (39.6 MB/s) - ‘hound.txt’ saved [345659/345659]



#### To find the number of Lines, Word Count and Characters we use wc

In [17]:
!wc hound.txt

  7224  62154 345659 hound.txt


In [7]:
!man grep

GREP(1)                          User Commands                         GREP(1)

NAME
       grep, egrep, fgrep, rgrep - print lines that match patterns

SYNOPSIS
       grep [OPTION...] PATTERNS [FILE...]
       grep [OPTION...] -e PATTERNS ... [FILE...]
       grep [OPTION...] -f PATTERN_FILE ... [FILE...]

DESCRIPTION
       grep  searches  for  PATTERNS  in  each  FILE.  PATTERNS is one or more
       patterns separated by newline characters, and  grep  prints  each  line
       that  matches a pattern.  Typically PATTERNS should be quoted when grep
       is used in a shell command.

       A FILE of “-”  stands  for  standard  input.   If  no  FILE  is  given,
       recursive  searches  examine  the  working  directory, and nonrecursive
       searches read standard input.

       In addition, the variant programs egrep, fgrep and rgrep are  the  same
       as  grep -E,  grep -F,  and  grep -r, respectively.  These variants are
       deprecated, but are provided for backward co

#### This is done to show that there are instances where Holmes's appears even after using the !grep -w.

In [2]:
!grep -w "Holmes" hound.txt

Chapter 1. Mr. Sherlock Holmes
Mr. Sherlock Holmes, who was usually very late in the mornings, save
Holmes was sitting with his back to me, and I had given him no sign of
“Good!” said Holmes. “Excellent!”
“Perfectly sound!” said Holmes.
“Really, Watson, you excel yourself,” said Holmes, pushing back his
I laughed incredulously as Sherlock Holmes leaned back in his settee and
“No mention of that local hunt, Watson,” said Holmes with a mischievous
of science, ask of Sherlock Holmes, the specialist in crime? Come in!”
entered his eyes fell upon the stick in Holmes’s hand, and he ran
“A presentation, I see,” said Holmes.
“Dear, dear, that’s bad!” said Holmes, shaking his head.
“Come, come, we are not so far wrong, after all,” said Holmes. “And now,
“A dabbler in science, Mr. Holmes, a picker up of shells on the shores
of the great unknown ocean. I presume that it is Mr. Sherlock Holmes
with that of your friend. You interest me very much, Mr. Holmes. I
Sherlock Holmes waved our strange visi

#### grep -w is used to find the exact match of Holmes. But it includes Holmes's along with Holmes. Similarly it is showing for other chracters in the play.

In [10]:
!grep -w "Holmes" hound.txt|wc -l
!grep -w "Watson" hound.txt| wc -l
!grep -w "Barrymore" hound.txt| wc -l
!grep -w "Mortimer" hound.txt| wc -l

190
113
72
89


#### Below logic is used to show in case there any plurals of the characters name present in the play

In [12]:
!grep -w "Holmess" hound.txt|wc -l
!grep -w "Barrymores" hound.txt| wc -l
!grep -w "Watsons" hound.txt| wc -l
!grep -w "Mortimers" hound.txt| wc -l

0
6
0
1


#### The following observations were made:
* The Hound Play consists of 7224 lines
* Holmes appeared apprx 190 times
* Watson appeared apprx 113 times
* Barrymore appeared apprx 72 times
* Mortimer appeared apprx 89 times
* The character Holmes appears maximum number of times in the text which signifies that he is the most famous among all the listed characters of the play.

#### Observations related to the methology of finding the above conclusion:
* While searching for the exact match of the text characters, for example:"Holmes", we used grep -w. According to me this worked best. Although in the output we are getting "Holmes's" included in the 190. There can be a possibility that the data is not 100% clean.
* There are instances of characters where the names of characters are in plural. For example: Barrymore v/s Barrymores. So I have shown the two variations separately for each character.


    

### Part B. Characters in Hamlet 
Use the text available at https://s3.amazonaws.com/dmfa-2020/project-1/hamlet.txt for this part.
How many times do each of the following characters have speaking lines in Hamlet? Keep in
mind that this is the text of a play.
Hamlet, Polonius, Ophelia, and Horatio

In [9]:
!wget https://s3.amazonaws.com/dmfa-2020/project-1/hamlet.txt

--2022-09-20 00:51:07--  https://s3.amazonaws.com/dmfa-2020/project-1/hamlet.txt
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.217.45.150
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.217.45.150|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 204885 (200K) [text/plain]
Saving to: ‘hamlet.txt’

hamlet.txt          100%[===================>] 200.08K  --.-KB/s    in 0.004s  

2022-09-20 00:51:07 (48.7 MB/s) - ‘hamlet.txt’ saved [204885/204885]



In [3]:
!wc  hamlet.txt

  5074  35064 204885 hamlet.txt


#### The output shows extra line not spoken by Hamlet. Hence this method does not provide the accurate output

In [5]:
!grep -E "^HAMLET." hamlet.txt

HAMLET, Prince of Denmark. CLAUDIUS, King of Denmark, Hamlet’s uncle.
HAMLET. [_Aside._] A little more than kin, and less than kind.
HAMLET. Not so, my lord, I am too much i’ the sun.
HAMLET. Ay, madam, it is common.
HAMLET. Seems, madam! Nay, it is; I know not seems. ’Tis not alone my
HAMLET. I shall in all my best obey you, madam.
HAMLET. O that this too too solid flesh would melt, Thaw, and resolve
HAMLET. I am glad to see you well: Horatio, or I do forget myself.
HAMLET. Sir, my good friend; I’ll change that name with you: And what
HAMLET. I am very glad to see you.—Good even, sir.— But what, in faith,
HAMLET. I would not hear your enemy say so; Nor shall you do my ear
HAMLET. I prithee do not mock me, fellow-student. I think it was to see
HAMLET. Thrift, thrift, Horatio! The funeral bak’d meats Did coldly
HAMLET. In my mind’s eye, Horatio.
HAMLET. He was a man, take him for all in all, I shall not look upon
HAMLET. Saw? Who?
HAMLET. The King my father!
HAMLET. For God’s love let m

In [8]:
!grep -E "^HAMLET." hamlet.txt | wc -l
!grep -E "^POLONIUS." hamlet.txt|wc -l
!grep -E "^OPHELIA." hamlet.txt|wc -l
!grep -E "^HORATIO." hamlet.txt | wc -l

359
86
58
110


#### Using the below logic we were able to eliminate the extra lines and displayed only lines spoken by the mentioned characters of the play

In [2]:
!grep "^HAMLET\." hamlet.txt

HAMLET. [_Aside._] A little more than kin, and less than kind.
HAMLET. Not so, my lord, I am too much i’ the sun.
HAMLET. Ay, madam, it is common.
HAMLET. Seems, madam! Nay, it is; I know not seems. ’Tis not alone my
HAMLET. I shall in all my best obey you, madam.
HAMLET. O that this too too solid flesh would melt, Thaw, and resolve
HAMLET. I am glad to see you well: Horatio, or I do forget myself.
HAMLET. Sir, my good friend; I’ll change that name with you: And what
HAMLET. I am very glad to see you.—Good even, sir.— But what, in faith,
HAMLET. I would not hear your enemy say so; Nor shall you do my ear
HAMLET. I prithee do not mock me, fellow-student. I think it was to see
HAMLET. Thrift, thrift, Horatio! The funeral bak’d meats Did coldly
HAMLET. In my mind’s eye, Horatio.
HAMLET. He was a man, take him for all in all, I shall not look upon
HAMLET. Saw? Who?
HAMLET. The King my father!
HAMLET. For God’s love let me hear.
HAMLET. But where was this?
HAMLET. Did you not speak to it?
H

In [9]:
!grep  "^HAMLET\." hamlet.txt | wc -l
!grep  "^POLONIUS\." hamlet.txt| wc -l
!grep  "^OPHELIA\." hamlet.txt| wc -l
!grep  "^HORATIO\." hamlet.txt| wc -l

358
86
58
107


#### The following observations were made:
* The Hamlet Play consists of 5074 lines.
* HAMLET spoke 358 times in the play
* HORATIO spoke 107 times in the play
* POLONIUS spoke 86 times in the play
* OPHELIA spoke 58 times in the play
* We conclude that Hamlet spoke maximum number of times in the play as compare to other characters mentioned here followed by Horatio.

#### Observations related to the methology of finding the above conclusion:
* There are different ways of finding the lines spoken by each character. According to me, !grep "^HAMLET\." worked best since it provided only the lines starting with uppercase followed by a period with respect to each character.
* In another logic, for example with grep -E "^HAMLET", there appears certain instances where the lines not spoken by HAMLET also appeared as shown in the text above.



####  Question 2 Part A 
#### Unzipped the data and combined the two inflating CSV files using csvstack.A new file:"biketrip.csv" is created

In [13]:
!pwd

/home/ubuntu/notebooks


In [18]:
!wget https://s3.amazonaws.com/dmfa-2020/project-1/2018-capitalbikeshare-tripdata.zip

--2022-09-23 05:51:19--  https://s3.amazonaws.com/dmfa-2020/project-1/2018-capitalbikeshare-tripdata.zip
Resolving s3.amazonaws.com (s3.amazonaws.com)... 54.231.229.144
Connecting to s3.amazonaws.com (s3.amazonaws.com)|54.231.229.144|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18365732 (18M) [application/zip]
Saving to: ‘2018-capitalbikeshare-tripdata.zip’

2018-capitalbikesha 100%[===================>]  17.51M  --.-KB/s    in 0.1s    

2022-09-23 05:51:19 (129 MB/s) - ‘2018-capitalbikeshare-tripdata.zip’ saved [18365732/18365732]



#### Unzip the files 

In [19]:
!unzip 2018-capitalbikeshare-tripdata.zip

Archive:  2018-capitalbikeshare-tripdata.zip
  inflating: 201807-capitalbikeshare-tripdata.csv  
  inflating: 201808-capitalbikeshare-tripdata.csv  


#### Combining the two files using csvstack and a new files is created by the name of "biketrip.csv"

In [20]:
!csvstack 201807-capitalbikeshare-tripdata.csv 201808-capitalbikeshare-tripdata.csv >biketrip.csv

#### Counting the number of lines in all the files

In [ ]:
!wc -l 201807-capitalbikeshare-tripdata.csv
!wc -l 201808-capitalbikeshare-tripdata.csv
!wc -l biketrip.csv

404762 201807-capitalbikeshare-tripdata.csv
403867 201808-capitalbikeshare-tripdata.csv
808628 biketrip.csv


#### Comment: 
* It is shown that the two files were combined successfully. Since the new file contains the sum of all the
  lines present in the two files. biketrip.csv file contains huge data to derive a certain conclusion. 
  But it will contain few errors also. And we will check them using csvclean.

#### Checking the errors. But this does not ensure that the data is 100% clean

In [24]:
!csvclean biketrip.csv

No errors.


#### csvstat shows the details of each columns in a structured format.Here we have used the data of 40000 rows.

In [25]:
!head -40000 biketrip.csv | csvstat

  1. "Duration"

	Type of data:          Number
	Contains null values:  False
	Unique values:         5436
	Smallest value:        60
	Largest value:         84571
	Sum:                   54903940
	Mean:                  1372.632816
	Median:                781
	StDev:                 2592.576769
	Most common values:    316 (54x)
	                       269 (53x)
	                       377 (52x)
	                       507 (51x)
	                       401 (49x)

  2. "Start date"

	Type of data:          DateTime
	Contains null values:  False
	Unique values:         36019
	Smallest value:        2018-07-01 00:00:58
	Largest value:         2018-07-04 13:31:14
	Most common values:    2018-07-01 13:52:26 (4x)
	                       2018-07-02 08:26:19 (4x)
	                       2018-07-02 10:47:26 (4x)
	                       2018-07-02 17:07:16 (4x)
	                       2018-07-02 17:34:44 (4x)

  3. "End date"

	Type of data:          DateTime
	Contains null values:  False
	Uniqu

#### List the labels for the heading line

In [26]:
!csvcut -n biketrip.csv

  1: Duration
  2: Start date
  3: End date
  4: Start station number
  5: Start station
  6: End station number
  7: End station
  8: Bike number
  9: Member type


#### 10 Capital Bikeshare most popular departing stations in July and August 2018 by station number and station name

* Column 4 and 5 represents Start Station Number and Start Sation respectively.

In [28]:
!csvcut -c4,5 biketrip.csv | sort | uniq -c | sort -rn | head

  15474 31258,Lincoln Memorial
  13082 31623,Columbus Circle / Union Station
  11999 31247,Jefferson Dr & 14th St SW
  11671 31289,Henry Bacon Dr & Lincoln Memorial Circle NW
  11650 31288,4th St & Madison Dr NW
  10357 31248,Smithsonian-National Mall / Jefferson Dr & 12th St SW
   9554 31249,Jefferson Memorial
   8620 31290,17th St & Independence Ave SW
   8583 31200,Massachusetts Ave & Dupont Circle NW
   8569 31201,15th & P St NW


#### 10 Capital Bikeshare most popular destination stations in July and August 2018 by station number and station name

* Column 6 and 7 represents End Station Number and End Sation respectively

In [30]:
!csvcut -c6,7 biketrip.csv | sort | uniq -c | sort -rn | head

  15642 31258,Lincoln Memorial
  13635 31623,Columbus Circle / Union Station
  12135 31247,Jefferson Dr & 14th St SW
  11722 31289,Henry Bacon Dr & Lincoln Memorial Circle NW
  11555 31288,4th St & Madison Dr NW
  10693 31248,Smithsonian-National Mall / Jefferson Dr & 12th St SW
   9866 31249,Jefferson Memorial
   9141 31200,Massachusetts Ave & Dupont Circle NW
   8884 31201,15th & P St NW
   8640 31290,17th St & Independence Ave SW


#### 10 Capital Bikeshare most popular departing-destination stations in July and August 2018 by station number and station name

* Sorted the Start and End Station to get the 10 most popular departing and destination stations.

In [33]:
!csvcut -c5,7 biketrip.csv | sort | uniq -c | sort -rn | head 

   1633 Smithsonian-National Mall / Jefferson Dr & 12th St SW,Smithsonian-National Mall / Jefferson Dr & 12th St SW
   1553 Jefferson Dr & 14th St SW,Jefferson Dr & 14th St SW
   1521 Lincoln Memorial,Jefferson Memorial
   1441 4th St & Madison Dr NW,4th St & Madison Dr NW
   1327 Lincoln Memorial,Lincoln Memorial
   1231 Jefferson Dr & 14th St SW,Lincoln Memorial
   1225 17th St & Independence Ave SW,Lincoln Memorial
   1121 Ohio Dr & West Basin Dr SW / MLK & FDR Memorials,Ohio Dr & West Basin Dr SW / MLK & FDR Memorials
   1115 17th St & Independence Ave SW,17th St & Independence Ave SW
   1085 Henry Bacon Dr & Lincoln Memorial Circle NW,Henry Bacon Dr & Lincoln Memorial Circle NW
sort: write failed: 'standard output': Broken pipe
sort: write error


#### Comments from the above output:
* It is shown in the outcome that the Lincoln Memorial is the most popular departing as well the most popular destination station.
* Also the top 7 stations are same in both departure and destination stations when viewed separately.
* We cannot comment about the revenue generation from the Lincoln Memorial,since we need some more parameters for it.
* For the popular departing- destination pair,Smithsonian-National and Jefferson Dr & 12th St SW are the most widely used stations by the people in DC area
* Followed by the first pair, Jefferson Dr & 14th St SW is the second most commonly used stations for entry and exit


#### Part B 

#### Below 12 bikes were used most in trips departing from the Lincoln Memorial which is the most popular departing station

In [2]:
!csvgrep -c5 -r "Lincoln Memorial" biketrip.csv | csvcut -c4-5,8 |sort | uniq -c | sort -rn | head -12

     18 31258,Lincoln Memorial,W01311
     17 31258,Lincoln Memorial,W22553
     16 31258,Lincoln Memorial,W00919
     15 31289,Henry Bacon Dr & Lincoln Memorial Circle NW,W23250
     15 31289,Henry Bacon Dr & Lincoln Memorial Circle NW,W20874
     15 31289,Henry Bacon Dr & Lincoln Memorial Circle NW,W20318
     15 31258,Lincoln Memorial,W23074
     15 31258,Lincoln Memorial,W22567
     15 31258,Lincoln Memorial,W21882
     15 31258,Lincoln Memorial,W20527
     14 31289,Henry Bacon Dr & Lincoln Memorial Circle NW,W21798
     14 31289,Henry Bacon Dr & Lincoln Memorial Circle NW,W20594
sort: write failed: 'standard output': Broken pipe
sort: write error


* This is the final output using Station number

#### Note : Instead of the Station name, station number used because the Lincoln Memorial appears in another station name also. As shown above 

In [34]:
!csvgrep -c4 -r "31258" biketrip.csv | csvcut -c4-5,8 |sort | uniq -c | sort -rn | head -12

     18 31258,Lincoln Memorial,W01311
     17 31258,Lincoln Memorial,W22553
     16 31258,Lincoln Memorial,W00919
     15 31258,Lincoln Memorial,W23074
     15 31258,Lincoln Memorial,W22567
     15 31258,Lincoln Memorial,W21882
     15 31258,Lincoln Memorial,W20527
     14 31258,Lincoln Memorial,W23311
     14 31258,Lincoln Memorial,W23003
     14 31258,Lincoln Memorial,W22369
     14 31258,Lincoln Memorial,W21089
     14 31258,Lincoln Memorial,W20638
sort: write failed: 'standard output': Broken pipe
sort: write error


#### Also we can add $ to show only values for Lincoln Memorial

In [3]:
!csvgrep -c5 -r "Lincoln Memorial$" biketrip.csv | csvcut -c4-5,8 |sort | uniq -c | sort -rn | head -12

     18 31258,Lincoln Memorial,W01311
     17 31258,Lincoln Memorial,W22553
     16 31258,Lincoln Memorial,W00919
     15 31258,Lincoln Memorial,W23074
     15 31258,Lincoln Memorial,W22567
     15 31258,Lincoln Memorial,W21882
     15 31258,Lincoln Memorial,W20527
     14 31258,Lincoln Memorial,W23311
     14 31258,Lincoln Memorial,W23003
     14 31258,Lincoln Memorial,W22369
     14 31258,Lincoln Memorial,W21089
     14 31258,Lincoln Memorial,W20638
sort: write failed: 'standard output': Broken pipe
sort: write error


#### Below 12 bikes were used most in trips for destination from the Lincoln Memorial which is the most popular destination station

In [5]:
!csvgrep -c7 -r "Lincoln Memorial$" biketrip.csv | csvcut -c6-7,8 | sort | uniq -c | sort -rn | head -12

     17 31258,Lincoln Memorial,W22553
     17 31258,Lincoln Memorial,W01311
     16 31258,Lincoln Memorial,W21882
     16 31258,Lincoln Memorial,W20527
     16 31258,Lincoln Memorial,W00919
     15 31258,Lincoln Memorial,W22567
     15 31258,Lincoln Memorial,W22369
     14 31258,Lincoln Memorial,W23311
     14 31258,Lincoln Memorial,W23074
     14 31258,Lincoln Memorial,W23003
     14 31258,Lincoln Memorial,W22120
     14 31258,Lincoln Memorial,W21089
sort: write failed: 'standard output': Broken pipe
sort: write error


### Comments from the above output:
* Bike no. W01311 is the most commonly used while travelling from Lincoln Memorial. 
* For the destination, Bike no. W22553 & W01311 are most commonly used by the public at Lincoln Memorial.

#### Question3  – advanced csvkit command

In [1]:
!pwd

/home/ubuntu/notebooks


#### Uploaded file containing header

In [2]:
!wget https://www.fec.gov/files/bulk-downloads/data_dictionaries/cm_header_file.csv

--2022-09-23 07:57:40--  https://www.fec.gov/files/bulk-downloads/data_dictionaries/cm_header_file.csv
Resolving www.fec.gov (www.fec.gov)... 18.160.46.98, 18.160.46.102, 18.160.46.26, ...
Connecting to www.fec.gov (www.fec.gov)|18.160.46.98|:443... connected.
HTTP request sent, awaiting response... 302 Moved Temporarily
Location: https://cg-519a459a-0ea3-42c2-b7bc-fa1143481f74.s3-us-gov-west-1.amazonaws.com/bulk-downloads/data_dictionaries/cm_header_file.csv [following]
--2022-09-23 07:57:40--  https://cg-519a459a-0ea3-42c2-b7bc-fa1143481f74.s3-us-gov-west-1.amazonaws.com/bulk-downloads/data_dictionaries/cm_header_file.csv
Resolving cg-519a459a-0ea3-42c2-b7bc-fa1143481f74.s3-us-gov-west-1.amazonaws.com (cg-519a459a-0ea3-42c2-b7bc-fa1143481f74.s3-us-gov-west-1.amazonaws.com)... 108.175.50.164
Connecting to cg-519a459a-0ea3-42c2-b7bc-fa1143481f74.s3-us-gov-west-1.amazonaws.com (cg-519a459a-0ea3-42c2-b7bc-fa1143481f74.s3-us-gov-west-1.amazonaws.com)|108.175.50.164|:443... connected.
HTTP

#### Uploaded the zip file

In [3]:
!wget https://www.fec.gov/files/bulk-downloads/2022/cm22.zip

--2022-09-23 07:57:53--  https://www.fec.gov/files/bulk-downloads/2022/cm22.zip
Resolving www.fec.gov (www.fec.gov)... 18.160.46.56, 18.160.46.26, 18.160.46.102, ...
Connecting to www.fec.gov (www.fec.gov)|18.160.46.56|:443... connected.
HTTP request sent, awaiting response... 302 Moved Temporarily
Location: https://cg-519a459a-0ea3-42c2-b7bc-fa1143481f74.s3-us-gov-west-1.amazonaws.com/bulk-downloads/2022/cm22.zip [following]
--2022-09-23 07:57:53--  https://cg-519a459a-0ea3-42c2-b7bc-fa1143481f74.s3-us-gov-west-1.amazonaws.com/bulk-downloads/2022/cm22.zip
Resolving cg-519a459a-0ea3-42c2-b7bc-fa1143481f74.s3-us-gov-west-1.amazonaws.com (cg-519a459a-0ea3-42c2-b7bc-fa1143481f74.s3-us-gov-west-1.amazonaws.com)... 108.175.48.144
Connecting to cg-519a459a-0ea3-42c2-b7bc-fa1143481f74.s3-us-gov-west-1.amazonaws.com (cg-519a459a-0ea3-42c2-b7bc-fa1143481f74.s3-us-gov-west-1.amazonaws.com)|108.175.48.144|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 819828 (801K) [app

#### Unzip the above file

In [4]:
!unzip cm22.zip

Archive:  cm22.zip
  inflating: cm.txt                  


#### Since we do not need the zip, we will remove it using rm

In [5]:
!rm cm22.zip

#### Checking for errors. It is showing errors in txt file

In [6]:
!csvclean cm.txt

1650 errors logged to cm_err.csv
32719 rows were joined/reduced to 8070 rows after eliminating expected internal line breaks.


#### Using head to show the top 10 columns and rows

In [7]:
!head -10 cm.txt

C00000059|HALLMARK CARDS, INC. PAC (HALLPAC)|KLEIN, CASSIE MS.|2501 MCGEE, MD853||KANSAS CITY|MO|64108|B|Q|UNK|M|C|HALLMARK CARDS, INC.|
C00000422|AMERICAN MEDICAL ASSOCIATION POLITICAL ACTION COMMITTEE|WALKER, KEVIN MR.|25 MASSACHUSETTS AVE, NW|SUITE 600|WASHINGTON|DC|200017400|B|Q||M||ALABAMA MEDICAL PAC|
C00000489|D R I V E POLITICAL FUND CHAPTER 886|JERRY SIMS JR|3528 W RENO||OKLAHOMA CITY|OK|73107|U|N||Q|L||
C00000547|KANSAS MEDICAL SOCIETY POLITICAL ACTION COMMITTEE|C. RICHARD BONEBRAKE, M.D.|623 SW 10TH AVE||TOPEKA|KS|66612|U|Q|UNK|T|T||
C00000729|AMERICAN DENTAL ASSOCIATION POLITICAL ACTION COMMITTEE|FAIR, JULIAN H DR. III|1111 14TH STREET, NW|SUITE 1100|WASHINGTON|DC|200055627|B|Q|UNK|M|M||
C00000885|INTERNATIONAL UNION OF PAINTERS AND ALLIED TRADES POLITICAL ACTION TOGETHER POLITICAL COMMITTEE|SMITH, GREGG|7234 PARKWAY DRIVE||HANOVER|MD|21076|B|Q|UNK|M|L|INTERNATIONAL UNION OF PAINTERS AND ALLIED TRADES|
C00000901|BUILD POLITICAL ACTION COMMITTEE OF THE NATIONAL ASSOCIATION O

#### Converting the txt file into csv

In [8]:
!csvformat -d "|" cm.txt > cm.csv

#### Again checking for errors. It shows none but the data is not 100% clean

In [9]:
!csvclean cm.csv

No errors.


In [10]:
!head -10 cm.csv

C00000059,"HALLMARK CARDS, INC. PAC (HALLPAC)","KLEIN, CASSIE MS.","2501 MCGEE, MD853",,KANSAS CITY,MO,64108,B,Q,UNK,M,C,"HALLMARK CARDS, INC.",
C00000422,AMERICAN MEDICAL ASSOCIATION POLITICAL ACTION COMMITTEE,"WALKER, KEVIN MR.","25 MASSACHUSETTS AVE, NW",SUITE 600,WASHINGTON,DC,200017400,B,Q,,M,,ALABAMA MEDICAL PAC,
C00000489,D R I V E POLITICAL FUND CHAPTER 886,JERRY SIMS JR,3528 W RENO,,OKLAHOMA CITY,OK,73107,U,N,,Q,L,,
C00000547,KANSAS MEDICAL SOCIETY POLITICAL ACTION COMMITTEE,"C. RICHARD BONEBRAKE, M.D.",623 SW 10TH AVE,,TOPEKA,KS,66612,U,Q,UNK,T,T,,
C00000729,AMERICAN DENTAL ASSOCIATION POLITICAL ACTION COMMITTEE,"FAIR, JULIAN H DR. III","1111 14TH STREET, NW",SUITE 1100,WASHINGTON,DC,200055627,B,Q,UNK,M,M,,
C00000885,INTERNATIONAL UNION OF PAINTERS AND ALLIED TRADES POLITICAL ACTION TOGETHER POLITICAL COMMITTEE,"SMITH, GREGG",7234 PARKWAY DRIVE,,HANOVER,MD,21076,B,Q,UNK,M,L,INTERNATIONAL UNION OF PAINTERS AND ALLIED TRADES,
C00000901,BUILD POLITICAL ACTION COMMITTEE OF THE NA

#### Combining the two files and a new file is created "candiates.csv"

In [11]:
!csvstack cm_header_file.csv cm.csv > candidates.csv

#### Showing the labels below in the candidates.csv file

In [12]:
!csvcut -n candidates.csv

  1: CMTE_ID
  2: CMTE_NM
  3: TRES_NM
  4: CMTE_ST1
  5: CMTE_ST2
  6: CMTE_CITY
  7: CMTE_ST
  8: CMTE_ZIP
  9: CMTE_DSGN
 10: CMTE_TP
 11: CMTE_PTY_AFFILIATION
 12: CMTE_FILING_FREQ
 13: ORG_TP
 14: CONNECTED_ORG_NM
 15: CAND_ID


#### The file consists of 19285 lines

In [13]:
!wc -l candidates.csv

19285 candidates.csv


#### csvstat shows the details of each column

In [14]:
!csvstat candidates.csv

  1. "CMTE_ID"

	Type of data:          Text
	Contains null values:  False
	Unique values:         19284
	Longest value:         9 characters
	Most common values:    C00000422 (1x)
	                       C00000489 (1x)
	                       C00000547 (1x)
	                       C00000729 (1x)
	                       C00000885 (1x)

  2. "CMTE_NM"

	Type of data:          Text
	Contains null values:  True (excluded from calculations)
	Unique values:         19138
	Longest value:         178 characters
	Most common values:    COLLINS FOR CONGRESS (4x)
	                       SCHNEIDER FOR CONGRESS (3x)
	                       JAMES TAYLOR FOR CONGRESS (3x)
	                       HOLD THEM ACCOUNTABLE (3x)
	                       PAC THE COURT (3x)

  3. "TRES_NM"

	Type of data:          Text
	Contains null values:  True (excluded from calculations)
	Unique values:         13589
	Longest value:         73 characters
	Most common values:    None (780x)
	                       DATWYLE

#### Displaying first 7 columns and 15 rows of the combined file. 16 is taken because first row contains headings.

In [23]:
!head -16 candidates.csv | csvlook --max-columns 7

| CMTE_ID   | CMTE_NM                                                                                         | TRES_NM                    | CMTE_ST1                     | CMTE_ST2        | CMTE_CITY     | CMTE_ST | ... |
| --------- | ----------------------------------------------------------------------------------------------- | -------------------------- | ---------------------------- | --------------- | ------------- | ------- | --- |
| C00000422 | AMERICAN MEDICAL ASSOCIATION POLITICAL ACTION COMMITTEE                                         | WALKER, KEVIN MR.          | 25 MASSACHUSETTS AVE, NW     | SUITE 600       | WASHINGTON    | DC      | ... |
| C00000489 | D R I V E POLITICAL FUND CHAPTER 886                                                            | JERRY SIMS JR              | 3528 W RENO                  |                 | OKLAHOMA CITY | OK      | ... |
| C00000547 | KANSAS MEDICAL SOCIETY POLITICAL ACTION COMMITTEE                                               | 

#### Total candidates for Repulican Party

In [21]:
!csvgrep -c11 -r "REP" candidates.csv | wc -l

3595


#### List of the Republican Committees in an NY office. It is sorted in descending order. In this state and Party affiliation is included.

In [17]:
!csvgrep -c11  -r "REP" candidates.csv | csvcut -c2,6,7,11,14 | csvgrep -c3 -r "NY" | csvsort -c1 -r |head -11 | csvlook

| CMTE_NM                             | CMTE_CITY     | CMTE_ST | CMTE_PTY_AFFILIATION | CONNECTED_ORG_NM              |
| ----------------------------------- | ------------- | ------- | -------------------- | ----------------------------- |
| ZUMBLUSKAS 2020                     | NEW YORK      | NY      | REP                  |                               |
| ZMICH FOR CONGRESS                  | SHIRLEY       | NY      | REP                  |                               |
| ZELDIN FOR CONGRESS RECOUNT         | SHIRLEY       | NY      | REP                  |                               |
| ZELDIN FOR CONGRESS                 | SHIRLEY       | NY      | REP                  | ZELDIN VICTORY COMMITTEE 2020 |
| YURI DASHEVSKY FOR US CONGRESS 2022 | BROOKLYN      | NY      | REP                  |                               |
| VAN DE WATER FOR CONGRESS           | MILLBROOK     | NY      | REP                  |                               |
| TUCKER FOR CONGRESS           

#### Only Committe Name, City and Connected Organization is shown. Also, grep is used before csvcut but the out put is same as above when we compare the Committe Name, City and Connected organisation

In [22]:
!csvgrep -c11  -r "(?i)REP" candidates.csv |csvgrep -c7 -r "(?i)NY" |csvcut -c2,6,14 | csvsort -c1 -r |head -11 | csvlook

| CMTE_NM                             | CMTE_CITY     | CONNECTED_ORG_NM              |
| ----------------------------------- | ------------- | ----------------------------- |
| ZUMBLUSKAS 2020                     | NEW YORK      |                               |
| ZMICH FOR CONGRESS                  | SHIRLEY       |                               |
| ZELDIN FOR CONGRESS RECOUNT         | SHIRLEY       |                               |
| ZELDIN FOR CONGRESS                 | SHIRLEY       | ZELDIN VICTORY COMMITTEE 2020 |
| YURI DASHEVSKY FOR US CONGRESS 2022 | BROOKLYN      |                               |
| VAN DE WATER FOR CONGRESS           | MILLBROOK     |                               |
| TUCKER FOR CONGRESS                 | BRONX         |                               |
| TRISH BERGIN FOR CONGRESS           | EAST ISLIP    |                               |
| TOM REED PAC                        | CORNING       |                               |
| THE JANSON COMMITTEE OF HARLEM

#### Displaying all the number of times each committee appeared

In [18]:
!csvcut -c2 candidates.csv | sort | uniq -c | sort -rn 

      4 COLLINS FOR CONGRESS
      3 SCHNEIDER FOR CONGRESS
      3 PAC THE COURT
      3 JUAN FOR CONGRESS
      3 JAMES TAYLOR FOR CONGRESS
      3 HOLD THEM ACCOUNTABLE
      2 WORKING PEOPLE RISING
      2 WIRTH FOR CONGRESS
      2 WINTER FOR MONTANA
      2 WINNING FOR WOMEN INC.
      2 WESTERN MICHIGAN VICTORY FUND
      2 US CHAMBER OF COMMERCE
      2 URBINA FOR CONGRESS
      2 TUNNELL FOR CONGRESS
      2 TILLIS-NRSC COMMITTEE
      2 TERRANCE MORKEECH ABRAHAM 2024 FOR PRESIDENT
      2 TARKANIAN FOR CONGRESS
      2 TAKE BACK AMERICA PAC
      2 SWAIN FOR CONGRESS
      2 SULLIVAN FOR CONGRESS
      2 STUDENTS 4 DEMOCRACY
      2 STEVEN CASTELLANO FOR US CONGRESS
      2 STAND UP REPUBLIC INC
      2 SOUTH ASIANS VOTE 2020
      2 SOLIS FOR CONGRESS
      2 SOCAL HEALTH CARE COALITION A PROJECT OF SIXTEEN THIRTY FUND
      2 SINGH FOR SENATE
      2 SIERRA CLUB
      2 SHEPHERD FOR CONGRESS
      2 SERVE AMERICA VICTORY FUND
      2 SENTINEL PAC
      2 SCOTT FOR CONGRESS


#### Using head, the below output displays only the Committee that appeared more than 2 times

In [19]:
!csvcut -c2 candidates.csv | sort | uniq -c | sort -rn | head -6

      4 COLLINS FOR CONGRESS
      3 SCHNEIDER FOR CONGRESS
      3 PAC THE COURT
      3 JUAN FOR CONGRESS
      3 JAMES TAYLOR FOR CONGRESS
      3 HOLD THEM ACCOUNTABLE
sort: write failed: 'standard output': Broken pipe
sort: write error


#### Here is another approach of finding the Committee that appears more than 2 times. 

In [20]:
!csvcut -c2 candidates.csv | sort | uniq -c | sort -rn | awk 'int($1)>=3'

      4 COLLINS FOR CONGRESS
      3 SCHNEIDER FOR CONGRESS
      3 PAC THE COURT
      3 JUAN FOR CONGRESS
      3 JAMES TAYLOR FOR CONGRESS
      3 HOLD THEM ACCOUNTABLE


#### Comments:
* There are committees that appears more than 2 times. And these are the same committee names with different details like PO box etc.
* We need to investigate and look for some more data relevant to committee before removing them. So, at this time I would not recommend deleting any data.  
#### Acknowledge Assitance: I would like to thank Ms. Bagya Madhuwanti for intoducing me the awk command.